[REEF-1750] Enable YARN Node Labels

scan3 opened the pull request:
This addressed the issue by
  * Enables REEF Runtime YARN to use this feature.

JIRA:
  [REEF-1750](https://issues.apache.org/jira/browse/REEF-1750)

Pull Request:
  This closes #

bgchun commented:
@motus Node labeling is a long-awaited feature. Would you be interested in reviewing this PR for our GSoC student? We need help to test this PR in the production cluster. Thanks!

markusweimer commented:
This expands REEF's `EvaluatorRequest` API for Node labels, right? Can you explain how node labels shall be supported or emulated on resource managers that don't support them?

I am asking because it is already possible to use YARN node labels on YARN via the interface `o.a.r.runtime.yarn.driver.YarnContainerRequestHandler`, which REEF Drivers can use in lieu of the `EvaluatorRequestor` interface, if they know they are on YARN.

bgchun commented:
@craig1583 It would be good for you to share the overall design with the community. 
I know both YARN and Mesos support node labels. That's why you expanded `EvaluatorRequest`? 
How do you support node labels in the stand-alone runtime? Do other resource managers such as Kubernetes support node labels as well?

markusweimer commented:
+1 on sharing the design and plan, preferably on JIRA.

jwang98052 commented on lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/YarnJobSubmissionClient.java:
```diff
@@ -85,7 +86,8 @@
                           final SecurityTokenProvider tokenProvider,
-                          final YarnSubmissionParametersFileGenerator jobSubmissionParametersGenerator) {
+                          final YarnSubmissionParametersFileGenerator jobSubmissionParametersGenerator,
+                          final String nodeLabelExpression) {
```
Not a named parameter?

craig1583 replied:
Does java supports named parameter?

bgchun replied:
Tang named parameter

craig1583 commented:
Thank you for your reviews.
@markusweimer As @bgchun mentioned, I expanded `EvaluatorRequest` to provide an interface that supports both node labeling on YARN and node attributes on Mesos. I will share my overall design and plan on [REEF-1758](https://issues.apache.org/jira/browse/REEF-1758). So far, I’ve considered only YARN and Mesos, but I think supporting node labels in stand-alone runtime would be needed as well. I will have a think about that.

This [document](https://kubernetes.io/docs/concepts/configuration/assign-pod-node/#step-one-attach-label-to-the-node
) says Kubernetes also supports node labeling. I will examine whether other resource managers support it.

jwang98052 commented on lang/java/reef-common/src/main/java/org/apache/reef/driver/evaluator/EvaluatorRequest.java:
```diff
@@ -55,14 +56,24 @@
+                   final List<String> nodeNames,
+                   final List<String> rackNames,
+                   final String runtimeName,
+                   final String nodeLabelExpression) {
```
A request is for multiple nodes. Are all the nodes in the request share the same label?

motus replied:
They should. And we can probably create a unit test to verify that - at least locally and on YARN

craig1583 replied:
@motus Thanks!

jwang98052 replied:
should it be called nodeLabel? It is just a string, not an expression.

jwang98052 commented on lang/java/reef-common/src/main/java/org/apache/reef/driver/parameters/NodeLabelExpression.java:
```diff
@@ -0,0 +1,32 @@
+import org.apache.reef.tang.annotations.NamedParameter;
+
+/**
+ * Node label expression string.
```
Put more document to describe what is node label.

craig1583 replied:
Yes, it doesn't seem sufficient. I will put more.

jwang98052 commented on lang/java/reef-common/src/test/java/org/apache/reef/runtime/common/driver/api/ResourceRequestEventImplTest.java:
```diff
@@ -0,0 +1,44 @@
+
+import org.junit.Assert;
+import org.junit.Test;
+
```
You might want to have test to reach YarnJobSubmissionClient constructors, one with nodeLabel one without it.

craig1583 replied:
I will update it.

jwang98052 commented on lang/java/reef-common/src/main/java/org/apache/reef/driver/parameters/NodeLabelExpression.java:
```diff
@@ -0,0 +1,32 @@
+ * Node label expression string.
+ */
+@NamedParameter(doc="Node label expression")
+public final class NodeLabelExpression implements Name<String> {
```
If NodeLabel is what is called inside YARN, name it as NodeLabel would be more consistent. Expression make people think about something like regular expression. But it is really a string label only.

craig1583 replied:
I used `NodeLabelExpression` because it is the term YARN uses. But I think `NodeLabel` is more appropriate if people think so. Given that this parameter is for not only YARN but other RMs such as Mesos and Kubernetes, we don't have to follow YARN's term.

motus replied:
I would keep `NodeLabelExpression` and mention in the javadocs that on certain platforms the labels' syntax can be more elaborate than just plain string tags.

craig1583 replied:
Yes, a node label could be expressed as a delimited string or a JSON array according to the platform it is on. @jwang98052 How do you think of it?

tcNickolas commented:
@craig1583 After the second commit in this pull request AppVeyor build started to fail building in VS2015 and VS2017. Strangely, it looks like some problem building Java side during .NET build, but Travis built Java code successfully. Please investigate

craig1583 commented:
@tcNickolas A set of error messages start in [Line 619](https://ci.appveyor.com/project/ApacheSoftwareFoundation/reef/build/1662-master/job/bshi4yj2weqr8l5w#L619) saying it can't find protobuf 2.5.0.
How can I deal with this kind of problem?
`
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-antrun-plugin:1.7:run (generate-sources) on project reef-common: An Ant BuildException has occured: 2.5.0 expected, but protoc version was:
`
`
[ERROR] around Ant part ...<fail message="2.5.0 expected, but protoc version was: ${protoc.version}">... @ 7:77 in C:\projects\reef\lang\java\reef-common\target\antrun\build-main.xml
`

tcNickolas commented:
My guess is that we're using an earlier version of protobuf; pull request #1334 is going to update it to 2.1.0 but 2.5.0 is way beyond that. You have to figure out why your change requires 2.5.0 and remove that dependency

motus commented on lang/java/reef-common/src/main/java/org/apache/reef/driver/evaluator/EvaluatorRequest.java:
```diff
@@ -55,14 +56,24 @@
     this.runtimeName = runtimeName;
+    this.nodeLabelExpression = nodeLabelExpression;
   }
-
```
I want this empty line back 😄

craig1583 replied:
Oh sure. I will update.

motus commented:
@tcNickolas It looks like the problem we encounter on AppVeyor is related to #1334. @craig1583 could you please check?

craig1583 commented:
@motus OK, I will add the unit test. 
#1334 seems updating protobuf version on cs build, but in my case, the error occurs from java build. I tried to reproduce this error, but it passes all tests in my local environment.

craig1583 commented:
### 1. Bug fixed.
Filled out missing points on the control flow.

### 2. Unit Test - all passed.
Test environment was a cluster that consists of 3 nodes, one for master and two for slaves. One slave is labeled as "mylabel" and another one wasn't.

I configured yarn for node labeling testing(including enable node labeling, set capacity of each partition) manually, and carried out three cases.

1. Set label for the job, and request a resource without label.
2. No label for the job, but driver requests a resource with a specific label.
3. No label for the job, and one request is for labeled resource, and another is not.

All of these tests came with the expected results. #1, #2 retrieved a container from the labeled node and #3 retrieved a contaier from the labeled node and another from the node without label.

### 3. Excluded the test from the commit

I excuded the [test](https://github.com/craig1583/reef/tree/TestNodeLabel)
 from this commit because it is not reproducible. YARN configuration must be set before running the test and should be restored. However to do this, you have to restart running yarn client. I think it would be problematic if it affects other running jobs.

craig1583 commented:
You can attach node label on job or individual resource request.
For a job, bind named parameter `NodeLabelExpression` on driverConfiguration. It will set the label on every resource requests for the job.
For an individual evaluator request, you can set label like this way.
```java
EvaluatorRequest.newBuilder()
          .setNumber(1)
          .setMemory(500)
          .setNumberOfCores(1)
          .setNodeLabelExpression(NodeLabelTestDriver.this.nodeLabelExpression)
```

tcNickolas commented:
I don't like adding the feature without a test for it. A feature not covered with tests is an invitation to accidentally break it and not find out until a year later, like it happened with driver restart feature.

Is there any way at all to include this test (do you mean `ResourceRequestEventImplTest`?)? Make it run in sequence with others, not in parallel, and do changes to Yarn configuration before and after? Change Yarn configuration of all tests before executing them? At the very least  keep it in the code and add instructions to running it manually as part of release process.

Now that I think of it, we should probably also document running driver restart example and include it in release... @jwang98052 Thoughts?

craig1583 commented:
@tcNickolas [This](https://github.com/craig1583/reef/tree/TestNodeLabel/lang/java/reef-tests/src/test/java/org/apache/reef/tests/nodelabel) is the test for node labeling.

Of course I can provide node labeling example code separately with yarn cluster setup instruction. But storing and restoring all the states seems so complicated, in such cases like the test label is already using in the cluster.

jwang98052 commented on lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/YarnJobSubmissionClient.java:
```diff
@@ -85,7 +87,8 @@
                           final SecurityTokenProvider tokenProvider,
-                          final YarnSubmissionParametersFileGenerator jobSubmissionParametersGenerator) {
+                          final YarnSubmissionParametersFileGenerator jobSubmissionParametersGenerator,
+                          @Parameter(NodeLabelExpression.class) final String nodeLabelExpression) {
```
Is the label for request evaluator only? If yes, why add it to Job submission? It should be in Evaluator request only.

jwang98052 replied:
How do you pass the value of nodeLabelExpression to YarnJobSubmissionClient and where do you bind the value to this named parameter?

jwang98052 replied:
The change is for Java only or .Net as well? I see the change is added to bridge client code, but no other bridge related code.

craig1583 replied:
A user can label on the job submission to label entire evaluator requests for the job.
To label on the job, you can set node label on driverConfiguraion before run a driver.
```java
final Configuration driverConfiguration = DriverConfiguration.CONF
.set(NodeLabelTestConfiguration.NODE_LABEL_EXPRESSION, "mylabel")
...
.build();
```
`NODE_LABEL_EXPRESSION` is an optional parameter.

The change is for java only.

bgchun commented:
What value do I pass in job submission if I only want a subset of nodes with a label, which I will do with evaluator requests? Exposing node label to job submission is not consistent with the current design.

Sent from my iPhone

2017. 8. 4. 오후 11:30 scahn <notifications@github.com> 작성:

> @craig1583 commented on this pull request.
> 
> In lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/YarnJobSubmissionClient.java:
> 
> > @@ -85,7 +87,8 @@
>                            final ClasspathProvider classpath,
>                            final YarnProxyUser yarnProxyUser,
>                            final SecurityTokenProvider tokenProvider,
> -                          final YarnSubmissionParametersFileGenerator jobSubmissionParametersGenerator) {
> +                          final YarnSubmissionParametersFileGenerator jobSubmissionParametersGenerator,
> +                          @Parameter(NodeLabelExpression.class) final String nodeLabelExpression) {
> A user can label on the job submission to label entire evaluator requests for the job.
> To label on the job, you can set node label on driverConfiguraion before run a driver.
> 
> final Configuration driverConfiguration = DriverConfiguration.CONF
> .set(NodeLabelTestConfiguration.NODE_LABEL_EXPRESSION, "mylabel")
> ...
> .build();
> NODE_LABEL_EXPRESSION is an optional parameter.
> 
> The change is for java only.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

craig1583 commented:
@bgchun If you want to do so, you don't set the label on driverConfiguration(it is optional) and set the label on individual evaluator requests.
If you think so, I will discard that part.

bgchun commented:
@craig1583 We didn't expose resource specs in the job submission API. The specs of requested resources are encoded in the Driver logic. Let's make the API consistent by removing node label from the job submission API.

johnyangk commented:
Maybe we can use Hadoop miniclusters for the tests: https://hadoop.apache.org/docs/r2.8.0/hadoop-project-dist/hadoop-common/CLIMiniCluster.html

@scan3 I'll share the code we wrote to spin up and use HDFS miniclusters.

scan3 commented:
@johnyangk Thanks for sharing!
I have a question about CLIMiniCluster. The description in your link saying `Using the CLI MiniCluster, users can simply start and stop a single-node Hadoop cluster`. But I think we need more than two slave nodes to test node labeling. Is it available to use multiple nodes with CLIMiniCluster?

scan3 commented:
I added a yarn node labeling example on `reef-examples`.
you can run this with the command below.
```
yarn jar lang/java/reef-examples/target/reef-examples-0.16.0-SNAPSHOT-shaded.jar org.apache.reef.examples.yarnnodelabel.REEFYarnNodeLabelTest
```

johnyangk commented:
Not sure. We've only used a single node HDFS cluster.

scan3 commented:
@johnyangk It seems unavailable to label on the nodes in CLIMiniCluster. 

So I added a yarn node labeling test as a yarn test, rather than a unit test. You can run this by executing `bin/yarnnodelabeltest.sh`.

bgchun commented:
@scan3 Can you update this branch to resolve conflicts?

scan3 commented:
@bgchun I resolved conflicts and rebased my repository.

jooykim commented on lang/java/reef-tests/src/test/java/org/apache/reef/tests/yarnnodelabel/NodeLabelTestDriver.java:
```diff
@@ -0,0 +1,150 @@
+      final NodeId nodeid = NodeId.newInstance(host, port);
+      LOG.log(Level.INFO, "NodeId: " + nodeid);
+
+      YarnClient client = YarnClient.createYarnClient();
```
@scan3 Thanks for the work! I would like to make a comment. As a client using REEF, creating a `YarnClient` using the host + port information to check the node label of the allocated evaluator seems rather cumbersome. Moreover, the client must import an extra library to resolve the node label information this way, and this must be done for each type of resource manager (ex. Mesos) being used. May I suggest adding this code inside REEF, and provide a way of accessing the node label with the `allocatedEvaluator` instance? (ex. include it as a mapping for the node descriptor)

jooykim commented:
@scan3 Could you take a look at my comment? Thank you :)

bgchun commented:
@scan3 Did you also check if tests in REEF .NET pass? AppVeyor build failed.

scan3 commented:
@jooykim I implemented `Set<String> getNodeLabelExpressions()` on `EvaluatorDescriptor` class. I will push here after testing.

scan3 commented:
@bgchun I didn't change anything on .Net side. I will check whether it is a transient failure.

scan3 commented:
!rebuild

scan3 commented:
@jooykim Now you can get node labels from `EvaluatorDescriptor`.

example:
```java
Set<String> nodeLabelExpressions = allocatedEvaluator.getEvaluatorDescriptor().getNodeLabelExpressions();
```

bgchun commented on lang/java/reef-common/src/main/java/org/apache/reef/runtime/common/driver/evaluator/EvaluatorDescriptorImpl.java:
```diff
@@ -36,17 +38,28 @@
   private final int numberOfCores;
   private EvaluatorProcess process;
   private final String runtimeName;
+  private final Set<String> nodeLabelExpressions;
```
How about using Optional?

scan3 replied:
Ok. I will update.

bgchun replied:
Did you update?

scan3 replied:
I wanted to use empty optional object instead of null in the constructor but It doesn't work as I expected. I will dig it up.

bgchun commented on lang/java/reef-common/src/main/java/org/apache/reef/runtime/common/driver/resourcemanager/ResourceEvent.java:
```diff
@@ -58,4 +60,9 @@
+  /**
+   * @return node label expression of the resource
+   */
+  Optional<Set<String>> getNodeLabelExpressions();
```
You used Optional here. Please be consistent.

motus replied:
Also, why do we need to wrap `Set` into `Optional`? An empty set should be good enough if you ask me (both `Set` and `Optional` are monads, after all) :)

scan3 replied:
YARN returns `null` if there is no label on a node. For consistency, We'd better to use `Optional` rather than empty `Set`.

bgchun commented on lang/java/reef-runtime-yarn/src/main/java/org/apache/reef/runtime/yarn/driver/YarnContainerManager.java:
```diff
@@ -535,13 +537,28 @@ private void handleNewContainer(final Container container) {
       LOG.log(Level.FINEST, "Allocated Container: memory = {0}, core number = {1}",
           new Object[] {container.getResource().getMemory(), container.getResource().getVirtualCores()});
 
+      YarnClient client = YarnClient.createYarnClient();
```
This part is odd. You're creating a yarn client in the driver to get node label expressions. This doesn't sound good.

scan3 replied:
Unfortunately, YARN container does not support node label getter. I don't like this way either, but there seems to be no other way.

bgchun replied:
Can you add a comment saying TODO ISSUE-NUMBER: update this part when YARN supports a method to retrieve the node label expression of a container and create a JIRA issue? 
Make sure you follow the exact TODO format.

bgchun replied:
Can  you add TODO?

scan3 replied:
Sorry I missed it. I will update soon!

bgchun commented on lang/java/reef-runtime-yarn/src/main/java/org/apache/reef/runtime/yarn/driver/YarnResourceRequestHandler.java:
```diff
@@ -64,13 +64,22 @@ public synchronized void onNext(final ResourceRequestEvent resourceRequestEvent)
 
-    for (int i = 0; i < resourceRequestEvent.getResourceCount(); i++) {
-      containerRequests[i] = new AMRMClient.ContainerRequest(resource, nodes, racks, pri, relaxLocality);
+    if (nodeLabelExpression.equals("")) {
```
Can you move `if` inside the `for` loop and make this part more concise?

scan3 replied:
I will update it.

bgchun commented on lang/java/reef-tests/src/test/java/org/apache/reef/tests/yarnnodelabel/NodeLabelTest.java:
```diff
@@ -0,0 +1,77 @@
+  @Test
+  public void testNodeLabel() throws BindException, InjectionException {
+    if (!(testEnvironment instanceof YarnTestEnvironment)) {
+      return;
```
Can you log a warning message?

motus replied:
Actually, you should probably use `org.junit.Assume` to skip the unit test in non-YARN mode, e.g.
```java
    Assume.assumeTrue(
        "This test requires a YARN Resource Manager to connect to",
        Boolean.parseBoolean(System.getenv("REEF_TEST_YARN")));
```
I am not sure if checking for an environment variable is better than using `instanceof`, but you get the idea...

scan3 replied:
I will update it.

bgchun commented on lang/java/reef-tests/src/test/java/org/apache/reef/tests/yarnnodelabel/NodeLabelTest.java:
```diff
@@ -0,0 +1,77 @@
+import org.junit.Test;
+
+/**
+ * Tests whether Evaluator allocations requested with a given node label are fulfilled.
```
Be more specific about the test scenario.

scan3 replied:
I will describe the scenario more precisely.

bgchun commented on lang/java/reef-tests/src/test/java/org/apache/reef/tests/yarnnodelabel/NodeLabelTestDriver.java:
```diff
@@ -0,0 +1,121 @@
+    public void onNext(final StopTime stopTime) {
+      LOG.log(Level.INFO, "# of total default containers: {0}",
+          NodeLabelTestDriver.this.defaultContainerCount);
+      if (1 != NodeLabelTestDriver.this.defaultContainerCount) {
```
`NodeLabelTestDriver.this.defaultContainerCount != 1`

bgchun commented on lang/java/reef-tests/src/test/java/org/apache/reef/tests/yarnnodelabel/NodeLabelTestDriver.java:
```diff
@@ -0,0 +1,121 @@
+
+      LOG.log(Level.INFO, "# of total labeled containers: {0}",
+          NodeLabelTestDriver.this.labeledContainerCount);
+      if (1 != NodeLabelTestDriver.this.labeledContainerCount) {
```
`NodeLabelTestDriver.this.labeledContainerCount != 1`

bgchun commented:
@scan3 I left comments. Could you address them? 

Can you also update your PR description, which does not have enough information? It would be great if you also add descriptions on the tests included in the PR description.

motus commented on bin/yarnnodelabeltest.sh:
```diff
@@ -0,0 +1,31 @@
+DEPENDENCY_JAR=`echo $REEF_HOME/lang/java/reef-tests/target/reef-tests-*-test-jar-with-dependencies.jar`
+CLASSPATH=`yarn classpath`
+
+CMD="java -cp $YARN_CONF_DIR:$DEPENDENCY_JAR:$CLASSPATH org.junit.runner.JUnitCore org.apache.reef.tests.yarnnodelabel.NodeLabelTest $*" 
```
why do we need that script? I would rather fix the existing `runyarntests.sh` and add an extra command line to specify a particular unit test instead of `org.apache.reef.tests.AllTestsSuite`, just like we do in the PowerShell version

scan3 replied:
I isolated this test because this test needs a properly configured YARN cluster, differ from other tests.

motus replied:
@scan3 My point is that we should be able to run tests like
```bash
runyarntests.sh org.apache.reef.tests.yarnnodelabel.NodeLabelTest ...
```
instead of creating a new shell script just for one test suite.

scan3 replied:
@motus OK. Then, where should I put the instruction of the test?

scan3 replied:
I will move it onto PR description :)

motus commented on lang/java/reef-runtime-yarn/src/main/java/org/apache/reef/runtime/yarn/driver/YarnContainerManager.java:
```diff
@@ -535,13 +537,28 @@ private void handleNewContainer(final Container container) {
+      client.init(new YarnConfiguration());
+      client.start();
+
+      Set<String> nodeLabelExpressions = null;
```
make it `final`

scan3 replied:
It cannot be final. If you make it `final`, you can't initialize as `null` in the catch block.

scan3 commented on lang/java/reef-common/src/main/java/org/apache/reef/driver/evaluator/EvaluatorDescriptor.java:
```diff
@@ -49,4 +51,9 @@
+  /**
+   * @return node label expressions allocated to this Evaluator
+   */
+  Set<String> getNodeLabelExpressions();
```
No. We can set only one node label on `EvaluatorRequest` for YARN.

motus commented on lang/java/reef-runtime-yarn/src/main/java/org/apache/reef/runtime/yarn/driver/YarnContainerManager.java:
```diff
@@ -535,13 +537,28 @@ private void handleNewContainer(final Container container) {
+        e.printStackTrace();
+      } catch (IOException e) {
+        e.printStackTrace();
+      }
```
don't handle exceptions like that. instead, write something like:
```java
final Set<String> nodeLabelExpressions;
try {
  nodeLabelExpressions = client.getNodeToLabels().get(container.getNodeId());
} catch (final YarnException | IOException e) {
  LOG.log(Level.WARN, "Error getting labels for node: " + container.getNodeId(), e);
  nodeLabelExpressions = null; // or a new empty Set?
}
```
note:
   * `final` variable `nodeLabelExpressions`
   * `final` exception
   * use of `|` in catch
   * use of `Logger.log(Level, String, Throwable)` specialization instead of `.printStackTrace()`

scan3 replied:
Thank you! I will update it, except making `nodeLabelExpressions` as a `final` variable.

motus commented on lang/java/reef-tests/src/test/java/org/apache/reef/tests/yarnnodelabel/DummyTask.java:
```diff
@@ -0,0 +1,38 @@
+/**
+ * A dummy REEF Task.
+ */
+public class DummyTask implements Task {
```
Can't we reuse some of the existing tasks, e.g. the one from HelloREEF?

scan3 replied:
OK. I will remove `DummyTask` and use `HelloTask`.

motus commented on lang/java/reef-tests/src/test/java/org/apache/reef/tests/yarnnodelabel/NodeLabelTestDriver.java:
```diff
@@ -0,0 +1,121 @@
+          .setNumberOfCores(1)
+          .setNodeLabelExpression(NodeLabelTestDriver.this.nodeLabelExpression)
+          .build();
+      LOG.log(Level.INFO, "Requested Evaluator with node label: " + reqToMylabel.getNodeLabelExpression());
```
use `{0}` instead of string concatenation

motus commented on lang/java/reef-tests/src/test/java/org/apache/reef/tests/yarnnodelabel/NodeLabelTestDriver.java:
```diff
@@ -0,0 +1,121 @@
+      LOG.log(Level.INFO, "Evaluator Allocated: {0}", allocatedEvaluator);
+
+      Set<String> nodeLabelExpressions = allocatedEvaluator.getEvaluatorDescriptor().getNodeLabelExpressions();
+      LOG.log(Level.INFO, "Node Labels on this node: " + nodeLabelExpressions);
```
use `{0}`

motus commented on lang/java/reef-tests/src/test/java/org/apache/reef/tests/yarnnodelabel/NodeLabelTestDriver.java:
```diff
@@ -0,0 +1,121 @@
+    public void onNext(final AllocatedEvaluator allocatedEvaluator) {
+      LOG.log(Level.INFO, "Evaluator Allocated: {0}", allocatedEvaluator);
+
+      Set<String> nodeLabelExpressions = allocatedEvaluator.getEvaluatorDescriptor().getNodeLabelExpressions();
```
final

motus commented on lang/java/reef-webserver/src/test/java/org/apache/reef/webserver/TestAvroSerializerForHttp.java:
```diff
@@ -162,6 +163,11 @@ public int getNumberOfCores() {
+
+    @Override
+    public Set<String> getNodeLabelExpressions() {
+      return null;
```
shouldn't we return an empty set instead?

scan3 replied:
YARN returns `null` when there is no label on a node. I think we'd better to return `null` for consistency.

bgchun commented:
@motus Thanks. Keep them coming. :)

bgchun commented:
@scan3 Can you check the AppVeyor failure and report which test fails? Thanks.
(In PR https://github.com/apache/reef/pull/1365, there is no test failure.)

scan3 commented:
@bgchun Thank you for your review. 

`Org.Apache.REEF.Tests.Functional.IMRU.TestFailMapperTasksOnDispose.TestFailedMapperOnLocalRuntime` seems to be a transient failure.

scan3 commented:
@bgchun @motus Thank you for your review! I've addressed your comments and tested on my cluster. :)

bgchun commented on lang/java/reef-common/src/main/java/org/apache/reef/driver/evaluator/EvaluatorRequest.java:
```diff
@@ -196,6 +209,12 @@ private Builder(final EvaluatorRequest request) {
       }
     }
 
+    @SuppressWarnings("checkstyle:hiddenfield")
```
Why SuppressWarnings?

scan3 replied:
Every existed setters in `EvaluatorRequest.Builder` takes a parameter that has the same name as its local variable like below.
```java
public T setMemory(final int megaBytes) {
      this.megaBytes = megaBytes;
      return (T) this;
}
```
IMHO, previous contributor seems to have turned this check off, because it's not really that confusing. So I decided to follow these precedents.

bgchun commented on lang/java/reef-runtime-yarn/src/main/java/org/apache/reef/runtime/yarn/driver/YarnContainerManager.java:
```diff
@@ -535,13 +537,26 @@ private void handleNewContainer(final Container container) {
 
+      YarnClient client = YarnClient.createYarnClient();
+      client.init(new YarnConfiguration());
+      client.start();
```
Do you have to start the client?

scan3 replied:
Yes it appears so. When I tried without starting the client, it couldn't get the labels.
I will add a line stopping the client after retrieving labels.

bgchun commented on lang/java/reef-tests/src/test/java/org/apache/reef/tests/yarnnodelabel/NodeLabelTest.java:
```diff
@@ -0,0 +1,79 @@
+ * Before you run this test, you should configure your YARN cluster to use two partition, "default" and "mylabel".
+ * You can refer to https://issues.apache.org/jira/browse/REEF-1750.
+ *
+ * Application manager starts on the default partition. Then the NodeLabelTestDriver submits a Evaluator that labeled
```
submits an EvaluatorRequest labeled with "mylabel" and one labeled with "default"?

scan3 replied:
Thank you for your correction. Actually, it submits an EvaluatorRequest labeled with "mylabel" and one without label. I will revise it.

bgchun commented on lang/java/reef-tests/src/test/java/org/apache/reef/tests/yarnnodelabel/NodeLabelTest.java:
```diff
@@ -0,0 +1,79 @@
+
+/**
+ * Tests whether Evaluator allocations requested with a given node label are fulfilled.
+ * Before you run this test, you should configure your YARN cluster to use two partition, "default" and "mylabel".
```
Before running this test, configure your ...

scan3 replied:
Thank you!

bgchun commented on bin/runyarntests.sh:
```diff
@@ -22,6 +22,10 @@ export REEF_TEST_YARN=true
 CLASSPATH=`yarn classpath`
 
-CMD="java -cp $YARN_CONF_DIR:$DEPENDENCY_JAR:$CLASSPATH org.junit.runner.JUnitCore org.apache.reef.tests.AllTestsSuite $*" 
+if [ $1 = "org.apache.reef.tests.yarnnodelabel.NodeLabelTest" ]; then
```
@motus Are you happy with this change?

motus replied:
No, I am not 😄
@scan3 please don't reference specific tests in the generic script. I will open a separate JIRA ticket to update the script in a minute. The fix should look like this:
```bash
CMD="java -cp $YARN_CONF_DIR:$DEPENDENCY_JAR:$CLASSPATH org.junit.runner.JUnitCore ${*:-org.apache.reef.tests.AllTestsSuite}"
```
(or one can use more sophisticated command line options etc.)

I'll send the PR for the script soon.

scan3 replied:
@motus It seems cool. Thank you!

bgchun commented:
@scan3 I left more comments. Could you address them?
@motus Let me know when you're happy with this PR.

motus commented:
OK I am done with the bash script fix. Please take a look at PR #1366, and you can remove fixes to that script from your PR

bgchun commented:
@scan3 Can you update this PR? I merged #1366 from @motus.

scan3 commented on lang/java/reef-tests/src/test/java/org/apache/reef/tests/yarnnodelabel/NodeLabelTest.java:
```diff
@@ -0,0 +1,79 @@
+/**
+ * Tests whether Evaluator allocations requested with a given node label are fulfilled.
+ * Before you run this test, you should configure your YARN cluster to use two partition, "default" and "mylabel".
+ * You can refer to https://issues.apache.org/jira/browse/REEF-1750.
```
Thank you!

scan3 commented:
@bgchun @motus Thank you! I've addressed your new comments and rebased this PR after #1366.

bgchun commented:
@markusweimer Thanks for the comments. Which way do you prefer?

bgchun commented:
@scan3 Can you reflect Markus's comments?

motus commented:
Running unit tests now. Will do another round of reviews shortly.

bgchun commented:
@motus Do you have a preference? @markusweimer suggested the following: "either more generic in the common API (e.g. a generic Map<String, String> in resource requests OR to move this to the YARN specific APIs".

motus commented:
@bgchun I think the generic approach would be nicer. later we can have labels in local mode at least

scan3 commented on lang/java/reef-common/src/main/java/org/apache/reef/driver/evaluator/EvaluatorDescriptor.java:
```diff
@@ -49,4 +52,9 @@
   String getRuntimeName();
+
+  /**
+   * @return node label expressions allocated to this Evaluator
```
Thank you for your comment!
Actually it means labels set on a node where the Evaluator is running.
I will update it.

scan3 commented:
@bgchun @motus @markusweimer Thank you for sharing your ideas. I will reflect it soon.

scan3 commented:
@bgchun @motus @markusweimer Thank you for sharing your ideas. I will reflect it soon.

bgchun commented:
@scan3 Any update?

scan3 commented:
@bgchun Sorry, I've updated but not tested yet because of some trouble on my cluster.

scan3 commented:
@bgchun @markusweimer I have addressed your comments. Would you please look at the code once again? Thanks!

bgchun commented:
@markusweimer Did @scan3 address your concern?

markusweimer commented on lang/java/reef-common/src/main/java/org/apache/reef/runtime/common/driver/evaluator/EvaluatorDescriptorImpl.java:
```diff
@@ -80,4 +94,8 @@ public int getNumberOfCores() {
   }
+
+  public Map<String, String> getNodeLabelExpressions() {
+    return this.nodeLabelExpressions.get();
```
The use of `Optional` around a collection seems curious. Why is it needed here? Can't we just have an empty collection instead?

markusweimer replied:
Also, this is unsafe, as the caller can change the returned `Map`.

scan3 replied:
Thank you for your comment! I will return `ImmutableMap<String, String>` for all getters.

scan3 replied:
Thank you for your comment! `Optional` seems not necessary for Map. I will discard it.

markusweimer commented on lang/java/reef-common/src/main/java/org/apache/reef/runtime/common/driver/resourcemanager/ResourceEventImpl.java:
```diff
@@ -33,6 +35,7 @@
   private final Optional<Integer> virtualCores;
   private final Optional<String> rackName;
   private final String runtimeName;
+  private final Optional<Map<String, String>> nodeLabelExpressions;
```
Same as above: Why the `Optional` of a collection type?

markusweimer commented on lang/java/reef-common/src/main/java/org/apache/reef/runtime/common/driver/resourcemanager/ResourceEvent.java:
```diff
@@ -58,4 +60,9 @@
+  /**
+   * @return node label expression of the resource
+   */
+  Optional<Map<String, String>> getNodeLabelExpressions();
```
Why is this a `Map`? Please add documentation as to what key and value are in the Java Docs.

markusweimer commented on lang/java/reef-common/src/main/java/org/apache/reef/runtime/common/driver/resourcemanager/ResourceEventImpl.java:
```diff
@@ -74,6 +78,11 @@ public String getRuntimeName() {
 
+  @Override
+  public Optional<Map<String, String>> getNodeLabelExpressions() {
+    return nodeLabelExpressions;
```
This is potentially unsafe, as the caller of this API can subsequently change the returned `Optional` or `Map`.

scan3 commented:
@markusweimer
Thank you for your review! IMHO, current API is not suitable for resource managers other than YARN. I've uploaded an updated design doc on JIRA(https://issues.apache.org/jira/browse/REEF-1750) could you please review this?

scan3 commented:
@markusweimer I have addressed your comments. 
Now the API is provided as key-value pair. For YARN, I used a static final string as a key because YARN only support a single string node label.

```java
NodeLabelTestDriver.this.evaluatorRequestor.submit(EvaluatorRequest.newBuilder()
    .setNumber(1)
    .setMemory(64)
    .setNumberOfCores(1)
    .setNodeLabels(YarnUtilities.REEF_YARN_NODE_LABEL_EXPRESSION, "mylabel")
    .build());
```
Would you please look at the code once again? Thanks!

markusweimer commented on lang/java/reef-common/src/main/java/org/apache/reef/driver/evaluator/EvaluatorDescriptor.java:
```diff
@@ -49,4 +51,9 @@
+  /**
+   * @return node labels on this Evaluator
+   */
+  Map<String, String> getNodeLabels();
```
Please document why this is a `Map`, and what key and value represent.

markusweimer commented on lang/java/reef-common/src/main/java/org/apache/reef/driver/evaluator/EvaluatorRequest.java:
```diff
@@ -96,6 +100,15 @@ public static Builder newBuilder(final EvaluatorRequest request) {
+   * @return the node labels requested.
+   */
+  public Map<String, String> getNodeLabels() {
+    return this.nodeLabels;
```
What if the receiver of this instance starts to change it? After all, a `Map` is mutable and we return a reference here.

markusweimer commented on lang/java/reef-common/src/main/java/org/apache/reef/driver/evaluator/EvaluatorRequest.java:
```diff
@@ -171,6 +184,7 @@ public boolean getRelaxLocality() {
     private final List<String> rackNames = new ArrayList<>();
     private String runtimeName = "";
     private boolean relaxLocality = true; //if not set, default to true
+    private Map<String, String> nodeLabels = new HashMap<>();
```
`final`?

markusweimer commented on lang/java/reef-common/src/main/java/org/apache/reef/runtime/common/driver/api/ResourceRequestEvent.java:
```diff
@@ -35,6 +36,11 @@
   /**
+   * @return The node labels.
+   */
+  Map<String, String> getNodeLabels();
```
Same as everywhere: Please document why this is a `Map`, and what key and value represent.

markusweimer commented on lang/java/reef-common/src/main/java/org/apache/reef/runtime/common/driver/api/ResourceRequestEventImpl.java:
```diff
@@ -47,6 +50,12 @@ private ResourceRequestEventImpl(final Builder builder) {
+
+  @Override
+  public Map<String, String> getNodeLabels() {
+    return nodeLabels;
```
Same as above: This hands a reference to a mutable collection to the caller. Please explain if and why this is safe.

markusweimer commented on lang/java/reef-common/src/main/java/org/apache/reef/runtime/common/driver/resourcemanager/ResourceEvent.java:
```diff
@@ -58,4 +60,11 @@
+   * In YARN, key is a static constant string and value is the node label.
+   * In Mesos, these are required Attributes.
+   */
+  Map<String, String> getNodeLabels();
```
Please sort the comment such that the `@return` is at the bottom.

This is interesting. Maybe the correct name for this field isn't "Node labels", but "Scheduling Constraints" instead? If we have this `string,string` map, we could funnel all sorts of constraints through to the various runtimes, right?

DifferentSC commented:
@markusweimer Thanks a lot for your comments. I have discussed with the REEF folks in our lab and @scan3. Here's the summary.

* Adding a generic API in REEF is beneficial in that REEF developers can specify the constraints of the nodes without being aware of the APIs of each RM clients. I got the feedback from the REEF folks in SNU that it is quite annoying to use RM client APIs directly in REEF applications.
* However, I also agree that current key-value API could be misleading, and not general enough too. (in case that we support other RMs which specify node contrains in a way that cannot be expressed as the key-value map)
* I agree that the name "Scheduling Constraint" is better than "Node Labeling". @scan3 will change the name.

So, my suggestion is changing the current way of designating node constrains(key-value maps) into more general way. We can make a new interface, called `SchedulingContraint` and make the implementation classes like `YarnSchedulingConstraint` and `MesosSchedulingConstraint` for setting RM-specific constraints through their own builders. `EvaluatorRequestor` will get the `SchedulingConstraint` from `setSchedulingConstraint()` method.

@markusweimer Would you please let me know whether my answer is enough for your question and how you think about the new plan? Thanks a lot again for your comments!

markusweimer commented:
This sounds much better. One concern is that we should have a generic way of adding constraints such that a Driver doesn't have to know whether it asks for node labels on YARN or Mesos. Maybe we can have something like this?

```java
public interface SchedulingConstraintBuilder{

  /**
   * Require the allocated Evaluators to be on nodes with the given label
   */
  public SchedulingConstraintBuilder requireNodeLabel(final String nodeLabel);

  // we can add more scheduling constraints supported by all runtimes here over time

  /**
   * @return the set of scheduling constraints to be passed to the EvaluatorRequestor
   */
  public SchedulingConstraint build();
```

Then, we could have specific injectable implementations for YARN, Mesos, Local, ... which have at least these methods, but can expose runtime specific constraint builders.

WDYT of this plan?

bgchun commented:
@markusweimer I like the plan.
It looks like the design's finally getting settled down. :)

@scan3 Can you update this PR to reflect the comments from @markusweimer and @DifferentSC?

scan3 commented:
@markusweimer Thank you for your suggestion!
One remaining question is, what should REEF do in response to calling inappropriate API? (For example, calling `requireNodeLabel(final String nodeLabel)` on Mesos.)

I can imagine two paths:
1. Raise an exception.
2. Ignore.

@markusweimer Would you please let me know which way do you prefer?

markusweimer commented:
I was actually thinking that both YARN and Mesos could implement `SchedulingConstraintBuilder.requireNodeLabel`, just differently? If there are APIs only available in YARN, we should have a `YARNSchedulingConstraintBuilder` that inherits from `SchedulingConstraintBuilder` to add them. That way, a Driver only sees the APIs they can call without error.

scan3 commented:
@markusweimer Thank you for sharing your idea! I made a new PR that addresses your comments. Could you please review this?

markusweimer commented on lang/java/reef-common/src/main/java/org/apache/reef/driver/evaluator/EvaluatorRequest.java:
```diff
@@ -41,13 +41,14 @@
                    final List<String> nodeNames,
                    final List<String> rackNames) {
-    this(number, megaBytes, cores, nodeNames, rackNames, "");
+    this(number, megaBytes, cores, nodeNames, rackNames, "", null, true);
```
Do we need this constructors? They are only called from within the builder, right?

markusweimer commented on lang/java/reef-common/src/main/java/org/apache/reef/driver/evaluator/EvaluatorRequest.java:
```diff
@@ -56,16 +57,16 @@
                    final List<String> rackNames,
                    final String runtimeName) {
-    this(number, megaBytes, cores, nodeNames, rackNames, runtimeName, true);
+    this(number, megaBytes, cores, nodeNames, rackNames, runtimeName, null, true);
```
Same as above.

markusweimer commented on lang/java/reef-common/src/main/java/org/apache/reef/driver/evaluator/EvaluatorRequest.java:
```diff
@@ -56,16 +57,16 @@
+    this(number, megaBytes, cores, nodeNames, rackNames, runtimeName, null, true);
   }
 
-
```
Please remove this pure formatting change.

markusweimer commented on lang/java/reef-common/src/main/java/org/apache/reef/driver/evaluator/EvaluatorRequest.java:
```diff
@@ -196,6 +209,12 @@ private Builder(final EvaluatorRequest request) {
     }
 
+    @SuppressWarnings("checkstyle:hiddenfield")
+    public T setSchedulingConstraint(final SchedulingConstraint schedulingConstraint) {
```
This assumes that there is only one scheduling constraint. While this is the case now, would we want this to be the case in the future? If not, it might be better to have a `Set` of them and replace this method with `addSchedulingConstraint`.

markusweimer commented on lang/java/reef-common/src/main/java/org/apache/reef/driver/evaluator/SchedulingConstraint.java:
```diff
@@ -0,0 +1,37 @@
+@Public
+@DriverSide
+@Provided
+public interface SchedulingConstraint<T> {
```
Why the type parameter? Wouldn't we have multiple implementations of this?

markusweimer commented on lang/java/reef-common/src/main/java/org/apache/reef/runtime/common/driver/api/ResourceRequestEventImpl.java:
```diff
@@ -118,6 +127,15 @@ public Builder mergeFrom(final ResourceRequestEvent resourceRequestEvent) {
+    /**
+     * Set scheduling constraint.
+     */
+    public Builder setSchedulingConstraint(final SchedulingConstraint schedulingConstraint) {
```
Same as above. Maybe `addSchedulingConstraint`?

markusweimer commented on lang/java/reef-runtime-yarn/src/main/java/org/apache/reef/runtime/yarn/driver/YarnContainerManager.java:
```diff
@@ -535,13 +538,38 @@ private void handleNewContainer(final Container container) {
+      client.init(new YarnConfiguration());
+      client.start();
+
+      String nodeLabelExpressionsFromNode = null;
```
Make this `final` and don't initialize it, as it is assigned exactly once per code path below.

markusweimer commented on lang/java/reef-runtime-yarn/src/main/java/org/apache/reef/runtime/yarn/driver/YarnContainerManager.java:
```diff
@@ -535,13 +538,38 @@ private void handleNewContainer(final Container container) {
+          }
+        }
+      } catch (final YarnException | IOException e) {
+        LOG.log(Level.WARNING, "Error getting label.", e);
```
Please be more expressive in the log message here. What happens when we ignore these exceptions? Will the scheduling constraint be ignored? If so, print which constraint got ignored as part of the log.

markusweimer commented on lang/java/reef-runtime-yarn/src/main/java/org/apache/reef/runtime/yarn/driver/evaluator/YarnSchedulingConstraint.java:
```diff
@@ -0,0 +1,41 @@
+/**
+ * Scheduling constraint(YARN node label expression) for evaluator requests.
+ */
+public final class YarnSchedulingConstraint implements SchedulingConstraint<String> {
```
Does this class have to be `public`?

markusweimer commented on lang/java/reef-runtime-yarn/src/main/java/org/apache/reef/runtime/yarn/driver/evaluator/YarnSchedulingConstraintBuilder.java:
```diff
@@ -0,0 +1,49 @@
+/**
+ * YARN scheduling constraint builder for evaluator requests.
+ */
+public final class YarnSchedulingConstraintBuilder implements SchedulingConstraintBuilder{
```
Does this class have to be `public`?

markusweimer commented on lang/java/reef-runtime-yarn/src/main/java/org/apache/reef/runtime/yarn/driver/evaluator/package-info.java:
```diff
@@ -0,0 +1,22 @@
+ * under the License.
+ */
+/**
+ * Evaluator-side implementation of YARN.
```
This is in the `driver.evaluator` package. This suggests that it is the driver-side view of evaluators. The comment says otherwise. Please explain.

markusweimer commented:
@scan3 Last activity on this was a month ago. What's the status of this PR?

scan3 commented:
@markusweimer I am sorry but the progress is currently suspended because the semester has begun. May I close this issue and reopen it when I am available?

markusweimer commented:
@scan3 that would be great! Ideally, leave a note on the JIRA with a link to your work-in-progress branch. That way, we can't loose track of it.

scan3 commented:
@markusweimer Sure I will. Thank you for your comment!
